plugins/woocommerce/src/Internal/Admin/ProductForm/Form.php

```diff
@@ -0,0 +1,242 @@
+			'cards'    => self::get_cards(),
+			'sections' => self::get_sections(),
+		);
+	}
+
+	/**
+	 * Returns list of registered fields.
+	 *
+	 * @param array $sort_by key and order to sort by.
+	 * @return array list of registered fields.
+	 */
+	public static function get_fields( $sort_by = array(
+		'key'   => 'order',
+		'order' => 'asc',
+	) ) {
+		return self::get_items( 'field', 'Field', $sort_by );
+	}
+
+	/**
+	 * Returns list of registered cards.
+	 *
+	 * @param array $sort_by key and order to sort by.
+	 * @return array list of registered cards.
+	 */
+	public static function get_cards( $sort_by = array(
+		'key'   => 'order',
+		'order' => 'asc',
+	) ) {
+		return self::get_items( 'card', 'Card', $sort_by );
+	}
+
+	/**
+	 * Returns list of registered sections.
+	 *
+	 * @param array $sort_by key and order to sort by.
+	 * @return array list of registered sections.
+	 */
+	public static function get_sections( $sort_by = array(
+		'key'   => 'order',
+		'order' => 'asc',
+	) ) {
+		return self::get_items( 'section', 'Section', $sort_by );
+	}
+
+	/**
+	 * Returns list of registered items.
+	 *
+	 * @param string       $type Form component type.
+	 * @param class-string $class_name Class of component type.
+	 * @param array        $sort_by key and order to sort by.
+	 * @return array       list of registered items.
+	 */
+	private static function get_items( $type, $class_name, $sort_by = array(
+		'key'   => 'order',
+		'order' => 'asc',
+	) ) {
+		$item_list = self::${ 'form_' . $type . 's' };
+		$class     = 'Automattic\\WooCommerce\\Internal\\Admin\\ProductForm\\' . $class_name;
+		$items     = array_values( $item_list );
+		if ( method_exists( $class, 'sort' ) ) {
+			usort(
+				$items,
+				function ( $a, $b ) use ( $sort_by, $class ) {
+					return $class::sort( $a, $b, $sort_by );
+				}
+			);
+		}
+		return $items;
+	}
+
+	/**
+	 * Creates a new item.
+	 *
+	 * @param string       $type Form component type.
+	 * @param class-string $class_name Class of component type.
+	 * @param string       $id Item id.
+	 * @param string       $plugin_id Plugin id.
+	 * @param array        $args additional arguments for item.
+	 * @return Field|Card|Section|WP_Error New product form item or WP_Error.
+	 */
+	private static function create_item( $type, $class_name, $id, $plugin_id, $args ) {
+		$item_list = self::${ 'form_' . $type . 's' };
+		$class     = 'Automattic\\WooCommerce\\Internal\\Admin\\ProductForm\\' . $class_name;
+		if ( isset( $item_list[ $id ] ) ) {
+			return new WP_Error(
+				'wc_product_form_' . $type . '_duplicate_field_id',
+				sprintf(
+				/* translators: 1: Item type 2: Duplicate registered item id. */
+					esc_html__( 'You have attempted to register a duplicate form %1$s with WooCommerce Form: %2$s', 'woocommerce' ),
+					$type,
+					'`' . $id . '`'
+				)
+			);
+		}
+
+		$missing_arguments = method_exists( $class, 'get_missing_arguments' ) ? $class::get_missing_arguments( $args ) : array();
```

Comment:
This abstraction took me a while to follow it backwards.  I'm wondering if it might just be easier to do something like:

```
try {
  new $class( $args );
} catch {
  return new WP_Error(...);
}
```

And then handle any exceptions inside the class itself instead of the parent.

Reply:
Good suggestion, I made that change!